### DISCUSSION TITLE
The ability of replacing StructuredBuffer?

### DISCUSSION THREAD

**Discussion by laicasaane** [Antoshidza/NSprites#6] [Q&A]
Hi, once again, thank you for this brilliant library.

Currently I'm in need of making my project compatible with at least Android 10, so `StructuredBuffer` cannot be used because only Android 11 and latest support it. Could you point out to me important parts of the library and in the shader I have to change or any knowledge I should be aware of? Given that I don't know much about shader.

Thanks!

**Comment by Antoshidza** [comment]
Hello! I want to help you but I limitted with my knowledge about rendering (even when I've created sprite rendering system).

So here things that I can tell:
* `RenderArchetype` depend on using `ComputeBuffer`. This thing used to be filled with entities data. Then I use `MaterialPropertyBlock.SetBuffer(id, buffer)` and here we can use `GraphicsBuffer` as well as `ComputeBuffer`. But I can also tell that `GraphicsBuffer` has no `BeginWrite` so it can't be used inside job, but it can be filled with `NativeArray<T>` on main thread, so theoretically we can write a little bit of abstractions to make `RenderArchetype` more flexible in a way of loading data to GPU.
* `RenderArchetype` doesn't depend on exact type of buffer you use in shader
* I've seen some mentions about using `StructuredBuffer<T>` inside fragment shader and not using it inside vertex shader, so lower versions can be supported

I would recommend write a simple test mono / ecs system by looking how NSprites performs rendering. All hardcoded. To just figure out what setup of buffers and other things works for your case. Once we know what type of Buffers / etc we also need in NSprites, then we can think of how it can be integrated in.

**Reply by laicasaane** [reply]
Is `_pointersProperty` necessary for NSprites? Or it is there only because `StructuredBuffer` needs it? If I can remove `StructuredBuffer` I think `_pointersProperty` should be removed as well. Is there anything I need to know about this?

**Reply by laicasaane** [reply]
I'm thinking about replacing `StructuredBuffer` with `UNITY_DEFINE_INSTANCED_PROP` and `UNITY_ACCESS_INSTANCED_PROP` following the some GPU Instancing examples and tutorials. I will combine shader properties into `float4`s so their value can be set by `MaterialPropertyBlock.SetVector` instead of `.SetBuffer`.

**Reply by laicasaane** [reply]
This is the shader following those tutorials. But I don't know if it works 😂
https://gist.github.com/laicasaane/4aeba98de592bfa42932969c8d7dcacf

**Reply by Antoshidza** [reply]
> Is `_pointersProperty` necessary for NSprites? Or it is there only because `StructuredBuffer` needs it? If I can remove `StructuredBuffer` I think `_pointersProperty` should be removed as well. Is there anything I need to know about this?

Pointers property used to support `Reactive` and `Static` update modes. This is the internal data that needed to effectively update only changed data. You can avoid using this property by using only `EachUpdate` properties. It is the most unperformant way to sync data, though property pointers use it, so it won't dramatically effect the performance.

> I'm thinking about replacing `StructuredBuffer` with `UNITY_DEFINE_INSTANCED_PROP` and `UNITY_ACCESS_INSTANCED_PROP` following the some GPU Instancing examples and tutorials. I will combine shader properties into `float4`s so their value can be set by `MaterialPropertyBlock.SetVector` instead of `.SetBuffer`.

I can't figure out how combining properties to `float4` can be made, becaues such data structure isn't array by itself, so it can be used for not instanced rendering only, or for shared data.

> This is the shader following those tutorials. But I don't know if it works 😂 https://gist.github.com/laicasaane/4aeba98de592bfa42932969c8d7dcacf

I think this is how SRP batching work and it's for URP properties. Though can you check if it works with NSprites?

**Reply by laicasaane** [reply]
I kinda understand now the logic behind the use of `CommandBuffer`. It seems if I need to remove it, I have to rewrite NSprites completely, in other words, to make a completely new library.

**Reply by laicasaane** [reply]
I come to a conclusion that it's not worth the effort. I should just go with the hybrid workflow then.

**Reply by Antoshidza** [reply]
> I kinda understand now the logic behind the use of `CommandBuffer`. It seems if I need to remove it, I have to rewrite NSprites completely, in other words, to make a completely new library.

If you can't use `ComputeBuffer` then I want to ask you - will you even use buffers? Because if not, then I can't imagine instance rendering. However more details can let us undestand can we implement needed functionality in NSprites or not.
In the end we can implement some kind of `IRenderArchetype` to integrate any kind of data loading to NSprites and unbound it from instance rendering.

**Reply by laicasaane** [reply]
>  will you even use buffers

The sad answer is I don't even know about this. This is all new to me. The only thing I know is `StructuredBuffer` is not supported on Android 10 and below. So I'm finding a workaround for this.

**Reply by BerezkovN** [reply]
Could you provide a link to prove your statement? I'm pretty sure that they are supported

**Reply by Antoshidza** [reply]
> The only thing I know is `StructuredBuffer` is not supported on Android 10 and below. So I'm finding a workaround for this.

What you see in unity? Error / nothing rendered / all rendered purple? Want to understand because I'm going to search / ask on forum about this. I also need to undestand limitations of my work.

[Here](https://forum.unity.com/threads/how-structuredbuffer-can-be-used-in-shaders-for-android.532240/) I've found some converstion about using `StructuredBuffer<T>` on Android.

**Reply by laicasaane** [reply]
I'm sorry I didn't remember correctly the Android version of my test device, turn out it's Android 9, not 10.

> I've seen some mentions about using StructuredBuffer<T> inside fragment shader and not using it inside vertex shader, so lower versions can be supported

I've moved them into the fragment shader but nothing renders on my device, same as before. There is just nothing: no purple texture, no exception.

Unfortunately, some minutes ago I've been told that our project must support Android 9. So I think I have to say goodbye to CommandBuffer. 😢

**Reply by Antoshidza** [reply]
> I've moved them into the fragment shader but nothing renders on my device, same as before. There is just nothing: no purple texture, no exception.

I think this is because `setup` function in shader also can't use `StructuredBuffer`, so matrix initialized wrong. You can try to set hardcoded matrix somewhere in the center of a scene and see if something renderers.

**Reply by laicasaane** [reply]
> You can try to set hardcoded matrix somewhere in the center of a scene and see if something renderers.

Could you give me an example? I don't understand how to do this really.

**Reply by Antoshidza** [reply]
Something like that
```hlsl
Shader "Universal Render Pipeline/2D/SimpleSpriteShader"
{
    Properties
    {
        _MainTex("_MainTex", 2D) = "white" {}
    }

    HLSLINCLUDE
    #include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/Core.hlsl"

    CBUFFER_START(UnityPerMaterial)
    CBUFFER_END
    ENDHLSL

    SubShader
    {
        Tags {"Queue" = "AlphaTest" "RenderType" = "TransparentCutout" "RenderPipeline" = "UniversalPipeline" }

        Pass
        {
            Tags { "LightMode" = "UniversalForward" "Queue" = "AlphaTest" "RenderType" = "TransparentCutout"}
            ZTest LEqual    //Default
            // ZTest Less | Greater | GEqual | Equal | NotEqual | Always
            ZWrite On       //Default
            Cull Off

            HLSLPROGRAM
            #pragma vertex UnlitVertex
            #pragma fragment UnlitFragment

            #pragma target 4.5
            #pragma exclude_renderers gles gles3 glcore
            #pragma multi_compile_instancing
            #pragma instancing_options procedural:setup

            struct Attributes
            {
                float3 positionOS   : POSITION;
                float2 uv			: TEXCOORD0;
                UNITY_VERTEX_INPUT_INSTANCE_ID
            };

            struct Varyings
            {
                float4  positionCS		: SV_POSITION;
                float2	uv				: TEXCOORD0;
                UNITY_VERTEX_INPUT_INSTANCE_ID
            };

            TEXTURE2D(_MainTex);
            SAMPLER(sampler_MainTex);

#if defined(UNITY_INSTANCING_ENABLED) || defined(UNITY_PROCEDURAL_INSTANCING_ENABLED) || defined(UNITY_STEREO_INSTANCING_ENABLED)
            StructuredBuffer<int> _propertyPointers;
            StructuredBuffer<float4> _mainTexSTBuffer;
            StructuredBuffer<int> _sortingIndexBuffer;
            StructuredBuffer<float4> _colorBuffer;
#endif

            void setup()
            {
                unity_ObjectToWorld = half4x4
                (
                    1, 0, 0, 0,
                    0, 1, 0, 0,
                    0, 0, 1, 0,
                    0, 0, 0, 1
                );
            }

            float2 TilingAndOffset(float2 UV, float2 Tiling, float2 Offset)
            {
                return UV * Tiling + Offset;
            }
            Varyings UnlitVertex(Attributes attributes, uint instanceID : SV_InstanceID)
            {
                Varyings varyings = (Varyings)0;

                UNITY_SETUP_INSTANCE_ID(attributes);
                UNITY_TRANSFER_INSTANCE_ID(attributes, varyings);

                varyings.positionCS = TransformObjectToHClip(attributes.positionOS);
                varyings.uv = attributes.uv;

                return varyings;
            }

            half4 UnlitFragment(Varyings varyings, uint instanceID : SV_InstanceID) : SV_Target
            {
#if defined(UNITY_INSTANCING_ENABLED) || defined(UNITY_PROCEDURAL_INSTANCING_ENABLED) || defined(UNITY_STEREO_INSTANCING_ENABLED)
                int propertyIndex = _propertyPointers[instanceID];
                float4 color = _colorBuffer[propertyIndex];
                float4 mainTexST = _mainTexSTBuffer[propertyIndex];
#else
                float4 color = float4(1,1,1,1);
                float4 mainTexST = float4(1,1,0,0);
#endif

                float2 uv = TilingAndOffset(varyings.uv, mainTexST.xy, mainTexST.zw);
                half4 texColor = SAMPLE_TEXTURE2D(_MainTex, sampler_MainTex, varyings.uv);
                clip(texColor.w - 0.5);

                return texColor * color;
            }
            ENDHLSL
        }
    }

    Fallback "Sprites/Default"
}
```

**Reply by laicasaane** [reply]
This doesn't work as well. On Unity editor it can render the sprites at the center of the game view. I've tested in development mode, there is no exception, but the entities are spawned correctly, and other systems work well too.
![image](https://user-images.githubusercontent.com/1594982/222220134-66e6e415-b82d-4a45-8547-a2c1457e3d4c.png)

**Reply by Antoshidza** [reply]
> This doesn't work as well. On Unity editor it can render the sprites at the center of the game view.

This was an idea of shader I attached to my previous reply, this version perform constant matrix initialization so any instance will ignore entity position component and just draw itself at zero position. 

> I've tested in development mode, there is no exception, but the entities are spawned correctly, and other systems work well too.

I'm a little bit confused. On screen sprites rendered as expected, and it happens in build with OpenGL ES 3.1 which in our case is a source of problems. So what you used to achieve this result on screen?

**Reply by laicasaane** [reply]
Sorry for the confusion, I should explain more clearly.

1. I've modified the shader according to your suggestions, there are 2:
    - (a) move `StructuredBuffer` to the fragment shader
    - (b) use a hardcoded matrix in the `setup` function
3. On Unity Editor: the sprites are rendered on the screen, as expected.
4. On Android 9 device: nothing is rendered.
5. On Android 11 device: the sprites appear on the screen too, same as on Unity Editor.
6. I've made a build in development mode to see if there is any exception or error reported on the Android 9 device. But there is none.

This Android 9 device says that it supports OpenGL ES 3.1 and Compute Shader. But nothing shows on its screen.

**Reply by Antoshidza** [reply]
I'm an idiot, sorry. There is `#pragma exclude_renderers gles gles3 glcore` which means it will exclude OpenGL ES 2.0 / OpenGL ES 3.x / OpenGL 3.x/4,x. So your device using OpenGL ES 3.1, so it wont render anything because of `#pragma exclude_renderers gles3 glcore` part.
I was writing this shader by tutorial of writing instanced shader. I don't remember why those renderers excluded, but you can try to remove this code part.

**Reply by laicasaane** [reply]
Many thanks! Remove that `#pragma` line does help!
Sprites are now rendered on my Android 9, but somehow incorrect. 😂

Here is a screenshot from the Android 9 device:
https://drive.google.com/file/d/1iGMF6vYwa13SqI-zecywTDtE1gNArflo/view?usp=share_link

Here is from the Android 11 device:
https://drive.google.com/file/d/1DIDFBBHq2NXkl_74IcSYxvn7vRDaFvDC/view?usp=share_link

**Reply by laicasaane** [reply]
The cause is these 2 lines. If these 2 line are commented, the issue goes away.

```
attributes.uv.x = flipValue.x >= 0 ? attributes.uv.x : (1.0 - attributes.uv.x);
attributes.uv.y = flipValue.y >= 0 ? attributes.uv.y : (1.0 - attributes.uv.y);
```

**Reply by Antoshidza** [reply]
> The cause is these 2 lines. If these 2 line are commented, the issue goes away.
> 
> ```
> attributes.uv.x = flipValue.x >= 0 ? attributes.uv.x : (1.0 - attributes.uv.x);
> attributes.uv.y = flipValue.y >= 0 ? attributes.uv.y : (1.0 - attributes.uv.y);
> ```

Maybe you using no flip component and NSprites uses default values, which can result in such unexpected rendering. Anyway, it seems we've fixed the problem :)

**Reply by laicasaane** [reply]
It' strange, I do use Flip component with value (1,1). This only happens on the Android 9 device which supports OpenGL ES 3.1. For devices support Vulkan 1.1, the flip components are correctly set.

**Reply by Antoshidza** [reply]
> It' strange, I do use Flip component with value (1,1). This only happens on the Android 9 device which supports OpenGL ES 3.1. For devices support Vulkan 1.1, the flip components are correctly set.

Try to do this operation in fragment shader. I have experience of wrong values, when passing values from vertex to fragment shader (AFAIK happens because values lerped between vertecies)

**Reply by laicasaane** [reply]
> Try to do this operation in fragment shader.

This solution doesn't work. 😞

**Reply by Antoshidza** [reply]
> > Try to do this operation in fragment shader.
> 
> This solution doesn't work. 😞

Can you please share rendering result and how shader code looks like?

**Reply by laicasaane** [reply]
> Can you please share rendering result

I was not being clear. "Doesn't work" means after the flip code is moved to the fragment shader, the sprites are still flickering up and down as seen in the screenshot above. And this issue just happens on the OpenGL ES 3.1 device, not the Vulkan 1.1 devices.

> how shader code looks like?

<details>
<summary>Yeah, here is the shader</summary>

```hlsl
Shader "NSprites/RegularNSprites_URP_Shader"
{
    Properties
    {
        _MainTex("_MainTex", 2D) = "white" {}
    }

    HLSLINCLUDE
    #include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/Core.hlsl"

    CBUFFER_START(UnityPerMaterial)
    CBUFFER_END
    ENDHLSL

    SubShader
    {
        Tags
        {
            "Queue" = "AlphaTest"
            "RenderType" = "TransparentCutout"
            "RenderPipeline" = "UniversalPipeline"
        }

        Pass
        {
            Tags
            {
                "LightMode" = "UniversalForward"
                "Queue" = "AlphaTest"
                "RenderType" = "TransparentCutout"
            }

            ZTest LEqual    //Default
            ZWrite On       //Default
            Cull Off

            HLSLPROGRAM

            #pragma vertex UnlitVertex
            #pragma fragment UnlitFragment

            #pragma target 4.5
            #pragma exclude_renderers gles
            #pragma multi_compile_instancing
            #pragma instancing_options procedural:setup

            struct Attributes
            {
                float3 positionOS   : POSITION;
                float2 uv			: TEXCOORD0;
                UNITY_VERTEX_INPUT_INSTANCE_ID
            };

            struct Varyings
            {
                float4  positionCS		: SV_POSITION;
                float2	uv				: TEXCOORD0;
                UNITY_VERTEX_INPUT_INSTANCE_ID
            };

            TEXTURE2D(_MainTex);
            SAMPLER(sampler_MainTex);

#if defined(UNITY_INSTANCING_ENABLED) || defined(UNITY_PROCEDURAL_INSTANCING_ENABLED) || defined(UNITY_STEREO_INSTANCING_ENABLED)
            StructuredBuffer<int> _propertyPointers;
            StructuredBuffer<float> _sortingValueBuffer;
            StructuredBuffer<float4> _mainTexSTBuffer;
            StructuredBuffer<float2> _positionBuffer;
            StructuredBuffer<float2> _pivotBuffer;
            StructuredBuffer<float2> _heightWidthBuffer;
            StructuredBuffer<int2> _flipBuffer;
#endif

            void setup()
            {
#if defined(UNITY_INSTANCING_ENABLED) || defined(UNITY_PROCEDURAL_INSTANCING_ENABLED) || defined(UNITY_STEREO_INSTANCING_ENABLED)
                int propertyIndex = _propertyPointers[unity_InstanceID];
                float2 scale = _heightWidthBuffer[propertyIndex];
                float2 renderPos = _positionBuffer[propertyIndex] - scale * _pivotBuffer[propertyIndex];
                unity_ObjectToWorld = half4x4
                (
                    scale.x, 0, 0, renderPos.x,
                    0, scale.y, 0, renderPos.y,
                    0, 0, 1, 0,
                    0, 0, 0, 1
                );
#endif
            }

            float2 TilingAndOffset(float2 UV, float2 Tiling, float2 Offset)
            {
                return UV * Tiling + Offset;
            }

            Varyings UnlitVertex(Attributes attributes, uint instanceID : SV_InstanceID)
            {
                Varyings varyings = (Varyings)0;

#if defined(UNITY_INSTANCING_ENABLED) || defined(UNITY_PROCEDURAL_INSTANCING_ENABLED) || defined(UNITY_STEREO_INSTANCING_ENABLED)
                int propertyIndex = _propertyPointers[instanceID];
                float4 mainTexST = _mainTexSTBuffer[propertyIndex];
                float sortingValue = _sortingValueBuffer[propertyIndex];
#else
                float4 mainTexST = float4(1, 1, 0, 0);
                float sortingValue = 0;
#endif

                UNITY_SETUP_INSTANCE_ID(attributes);
                UNITY_TRANSFER_INSTANCE_ID(attributes, varyings);

                varyings.positionCS = TransformObjectToHClip(attributes.positionOS);
                varyings.positionCS.z = sortingValue;
                varyings.uv = TilingAndOffset(attributes.uv, mainTexST.xy, mainTexST.zw);

                return varyings;
            }

            half4 UnlitFragment(Varyings varyings, uint instanceID : SV_InstanceID) : SV_Target
            {
#if defined(UNITY_INSTANCING_ENABLED) || defined(UNITY_PROCEDURAL_INSTANCING_ENABLED) || defined(UNITY_STEREO_INSTANCING_ENABLED)
                int propertyIndex = _propertyPointers[instanceID];
                int2 flipValue = _flipBuffer[propertyIndex];
#else
                int2 flipValue = int2(0, 0);
#endif

                varyings.uv.x = flipValue.x >= 0 ? varyings.uv.x : (1.0 - varyings.uv.x);
                varyings.uv.y = flipValue.y >= 0 ? varyings.uv.y : (1.0 - varyings.uv.y);

                half4 texColor = SAMPLE_TEXTURE2D(_MainTex, sampler_MainTex, varyings.uv);
                clip(texColor.w - 0.5);
                return texColor;
            }

            ENDHLSL
        }
    }

    Fallback "Sprites/Default"
}
```

</details>

**Reply by Antoshidza** [reply]
@laicasaane hmmm, looks wierd. I see few possible reasons of such behaviour:
* flip buffer values on entities are incorrect, so you can inspect them. But as I remember there was no system which writes to flip component
* fragment shader doesn't work with `StructuredBuffer` in your case

I would check the latter by next actions: Add some another instanced value, for example color. To do so you need and _colorBuffer to shader, access it's value in fragment shader. Add property data to properties set asset in project. Register component which should be treated as color (float4). And finally add this compont to entity with rand value on baking. Multiplay texColor with insanced color. See if color applyed to result.

Can you please remind me: flip operation also didn't work in vertex shader? So it just doesn't work while other things do?

**Reply by laicasaane** [reply]
My biggest suspect would be my OpenGL ES 3.1 device only supports a limited amount of `StructuredBuffer`s. Because when I add color to the shader as you suggest, the sprites are rendered black on this device.

I've also tested with a hardcoded flip value `int2(-1,1)` and the flip operation works correctly. So that's why I suspect my device's GPU doesn't like more `StructuredBuffer`.

**Reply by laicasaane** [reply]
I have an idea to test this theory: we can somehow calculate `renderPos` on the C# side, to remove scale and pivot buffers from the shader to leave space for flip buffer.

```
float2 renderPos = _positionBuffer[propertyIndex] - scale * _pivotBuffer[propertyIndex];
```

**Reply by Antoshidza** [reply]
@laicasaane 
* You can calculate on CPU side (c# in our case 😄) whatever you want and then pass result through buffer
* You can use custom struct with buffer, so you can encapsulate any set of data you want. Just ensure that component has exact same layout. ~Unfortunetelly NSprites can only operate with already known layouts. But I will see if this can be improved (the main problem is registering generic jobs of type we need to read from entities, and there is 4 generic jobs I use in `RenderArchetype` so alot boilerplate to give it to register to user).~ (NSprites supports any blittable data layout from 3.0,0 version)
* You can use `SystemInfo.maxComputeBufferInput` / `SystemInfo.maxComputeBufferInputFragment` / `SystemInfo.maxComputeBufferInputVertex` to see how much buffers you can read from your device.

**Reply by laicasaane** [reply]
My device reports:

```
SystemInfo.maxComputeBufferInput = 16
SystemInfo.maxComputeBufferInputFragment = 16
SystemInfo.maxComputeBufferInputVertex = 16
```

I've already implemented various test cases but no one works. The sprites are still flickering, both for flip and color. Even reducing the number of `StructuredBuffer` didn't work.

**Reply by laicasaane** [reply]
Another issue has just come to my attention. The sorting order is also wrong too! On the Android 9 device, sorting order is reversed. Sprites at higher Y position are rendered on top.

**Reply by Antoshidza** [reply]
> Another issue has just come to my attention. The sorting order is also wrong too! On the Android 9 device, sorting order is reversed. Sprites at higher Y position are rendered on top.

This one is tough but has clear ways to fix. You can replace changing `SV_POSITION` by for example shifting Z in matrix, so sprites will have depth depending on theirs sorting value.

I guess this issue exists because different devices has different Z rendering order. Exampled shader definitely not the most worked out, but it is just raw example :)

> I've already implemented various test cases but no one works. The sprites are still flickering, both for flip and color. Even reducing the number of `StructuredBuffer` didn't work.

The last thing I would try to do is just reading those values (color, flip) in vertex shader and then just pass to fragment shader by writing values to `Varyings` struct. If it wont help I would ask more expirienced guys on forum why such result can occur

**Reply by Antoshidza** [reply]
Also I want to remind you that any `Reactive` or `Static` properties should be accessed by property index rather than by instanceID. So firstly you should retrieve property index using instanceID like `int propIndex = _propertyIndex[instanceID];` and only then get color like `float4 color = _colorBuffer[propertyIndex];`

**Reply by laicasaane** [reply]
> float4 color = _colorBuffer[propertyIndex];

I do this actually. So it's not the source of the problem.

> The last thing I would try to do is just reading those values (color, flip) in vertex shader and then just pass to fragment shader by writing values to Varyings struct.

I do this for color, but the color still flickering between black and the actual color. So it won't help.

> This one is tough but has clear ways to fix. You can replace changing SV_POSITION by for example shifting Z in matrix, so sprites will have depth depending on theirs sorting value.

Could you please help me on this? 😂

**Reply by Antoshidza** [reply]
> > This one is tough but has clear ways to fix. You can replace changing SV_POSITION by for example shifting Z in matrix, so sprites will have depth depending on theirs sorting value.
> 
> Could you please help me on this? 😂

Instead of `varyings.positionCS.z = 1.0 / (sortingIndex + 1);` in vertex shader try do in `setup()` function
```hlsl
unity_ObjectToWorld = half4x4
(
    scale.x, 0, 0, renderPos.x,
    0, scale.y, 0, renderPos.y,
    0, 0, 1, 1.0 / (sortingIndex + 1),
    0, 0, 0, 1
);
```

**Reply by Antoshidza** [reply]
> I do this for color, but the color still flickering between black and the actual color. So it won't help.

Seems like buffer's values messed up. With position and other same between sprites values we have no problems, because messing them up leads to no visual difference. From another view your flip values also same between sprites (aren't they?) so messed up property indecies isn't an answer.

**Reply by Antoshidza** [reply]
@laicasaane while we was trying to find solvation, I've improved NSprties so now it can use any blittable component layout. So you can use now any count of blittable fields inside your component you use as properties.

**Reply by laicasaane** [reply]
> From another view your flip values also same between sprites (aren't they?) so messed up property indecies isn't an answer.

I don't really understand this part.

**Reply by Antoshidza** [reply]
> I don't really understand this part.

I was thinking about maybe values in buffers are messed up. If so we should see only color flickering because it is the only property which differ between entities, other values are the same (except position also, but messing this up leads to nothing, because other properties are the same, so messed position aren't visible).

**Reply by laicasaane** [reply]
This is the test with various sprites spawned at the same time. I don't know if this help you figure out something? (NSprites 2.3.0)
https://drive.google.com/file/d/1zmetdYE3NfeuMGbqoTYj0GKu0bQ7IKPs/view?usp=share_link

> except position also, but messing this up leads to nothing, because other properties are the same, so messed position aren't visible

I don't really understand the internal working of NSprites. But I guess you're updating buffers for sprites with the same renderId at once? So that messed up positions won't be visible?

**Reply by laicasaane** [reply]
Do you think that a combination of GPU Instancing shader (without `StructuredBuffer`), `Graphics.DrawMeshInstanced`, and `MaterialPropertyBlock` could work? I think we can sacrifice some performance to achieve broader compatibility.

**Reply by Antoshidza** [reply]
> This is the test with various sprites spawned at the same time. I don't know if this help you figure out something? (NSprites 2.3.0)
> https://drive.google.com/file/d/1zmetdYE3NfeuMGbqoTYj0GKu0bQ7IKPs/view?usp=share_link

Looks normal. If you use sprite atlasing, then all this sprites share same texture, which means they have different `MainTexST` component values and this values all correct (not messed up), so my assumption wrong, which means that data loaded to buffers in right sequence.
You can check Window -> Entities -> NSprites and look at RenderArchetype list, if you see only one archetype then I'm right.
Also did you remove flipping from shader or flickering just gone?

> Do you think that a combination of GPU Instancing shader (without `StructuredBuffer`), `Graphics.DrawMeshInstanced`, and `MaterialPropertyBlock` could work? I think we can sacrifice some performance to achieve broader compatibility.

How then you will read values of instances in shader without `StructuredBuffer`?

**Reply by laicasaane** [reply]
> Also did you remove flipping from shader or flickering just gone?

Yes, I removed it from the shader.

> I've improved NSprties so now it can use any blittable component layout. So you can use now any count of blittable fields inside your component you use as properties.

As I understand we can now reduce the number of `StructuredBuffer` in the shader, right?

As a side note, maybe I will put up another test with this shader to see if SRP Batching can work for sprites on Unity 2022.2. (Unity has made it works for Unity 2023.1 though). If the result is positive, I might switch to the hybrid solution (GameObject + Sprite Renderer + this shader) in the meantime.
https://gist.github.com/laicasaane/b879b3dddcc477e897666db8116fc95b

**Reply by laicasaane** [reply]
> You can check Window -> Entities -> NSprites and look at RenderArchetype list, if you see only one archetype then I'm right.

I've registered 61 sprite sheets and use 3 of them:
(screenshots captured in Unity Editor)

![image](https://user-images.githubusercontent.com/1594982/222968782-622136c1-5adf-4ffe-a75a-69cfa60a153e.png)
![image](https://user-images.githubusercontent.com/1594982/222968709-6947f341-359e-42e8-900d-96a3104eb3f3.png)

> If you use sprite atlasing

Yes, these sprites here are put into a single SpriteAtlas.

**Reply by Antoshidza** [reply]
> As I understand we can now reduce the number of `StructuredBuffer` in the shader, right?

Yep

> As a side note, maybe I will put up another test with this shader to see if SRP Batching can work for sprites on Unity 2022.2. (Unity has made it works for Unity 2023.1 though). If the result is positive, I might switch to the hybrid solution (GameObject + Sprite Renderer + this shader) in the meantime.
> https://gist.github.com/laicasaane/b879b3dddcc477e897666db8116fc95b

Would be nice to see SRP batching working with 2D sprites, because THIS is the source reason I even start to thinking about my own dots solution.

> I've registered 61 sprite sheets and use 3 of them:
> (screenshots captured in Unity Editor)
> Yes, these sprites here are put into a single SpriteAtlas.

Wow, so many render archetypes! If you use same material and all in the single atlas (which means 1 solid texture) then you can use just 1 archetype instead of registering 1 per sprite sheet. The trick is change UVs through `MainTexST` component (Foundation authorings can do it)

Well! What we have is - in test different sheets have different render archetypes, and all sprites have no flip and any additional problematic properties like color we was trying to test. This means that values can be messed up in buffer and in your case it won't be visible at all. Can you, please, put all of this different sprites into the same archetype?

**Reply by laicasaane** [reply]
> Wow, so many render archetypes! If you use same material and all in the single atlas (which means 1 solid texture) then you can use just 1 archetype instead of registering 1 per sprite sheet. The trick is change UVs through MainTexST component (Foundation authorings can do it)

Yes. Thanks to your foundation code I've learnt alot: my code is based on yours. But this is just a prototype so I can afford to have this many archetype. Optimization will come later when I have a concrete workflow for my project.

**Reply by Antoshidza** [reply]
> Yes. Thanks to your foundation code I've learnt alot: my code is based on yours. But this is just a prototype so I can afford to have this many archetype. Optimization will come later when I have a concrete workflow for my project.

Yes, I understand this. But rendering more then 1 visual object per render gives us possibility to understand what may be going wrong.
